_posts/2021-01-15-tuples-bring-generic-programming-to-scala-3.md

```diff
@@ -0,0 +1,326 @@
+is particularly interesting for our use case:
+
+> The compiler automatically generates instances of `Mirror` for `enum`s and
+> their cases, **case classes** and case objects, sealed classes or traits
+> having only case classes and case objects as children.
+
+That's why we can obtain a tuple from a case class using:
+```scala
+val bob: Employee = Employee("Bob", 42, false)
+val bobTuple: (String, Int, Boolean) = Tuple.fromProductTyped(bob)
+```
+But that is also why we can revert the operation:
+```scala
+val bobAgain: Employee = summon[Mirror.Of[Employee]].fromProduct(bobTuple)
+```
+
+# New tuples operations
+In the previous example, we saw that we can use `.head` and `.tail` on tuples,
+but Scala 3 introduces many other operations, here is a quick overview:
+
+| Operation  | Example                                                     | Result                                               |
+|------------|-------------------------------------------------------------|------------------------------------------------------|
+| `size`     | `(1, 2, 3).size`                                            | `3`                                                  |
+| `head`     | `(3 *: 4 *: 5 *: EmptyTuple).head`                          | `3`                                                  |
+| `tail`     | `(3 *: 4 *: 5 *: EmptyTuple).tail`                          | `(4, 5)`                                             |
+| `*:`       | `3 *: 4 *: 5 *: 6 *: EmptyTuple`                            | `(3, 4, 5, 6)`                                       |
+| `++`       | `(1, 2, 3) ++ (4, 5, 6)`                                    | `(1, 2, 3, 4, 5, 6)`                                 |
+| `drop`     | `(1, 2, 3).drop(2)`                                         | `(3)`                                                |
+| `take`     | `(1, 2, 3).take(2)`                                         | `(1, 2)`                                             |
+| `apply`    | `(1, 2, 3)(2)`                                              | `3`                                                  |
+| `splitAt`  | `(1, 2, 3, 4, 5).splitAt(2)`                                | `((1, 2), (3, 4, 5))`                                |
+| `zip`      | `(1, 2, 3).zip(('a', 'b'))`                                 | `((1 'a'), (2, 'b'))`                                |
+| `toList`   | `(1, 'a', 2).toList`                                        | `List(1, 'a', 2) : List[Int | Char]`                 |
+| `toArray`  | `(1, 'a', 2).toArray`                                       | `Array(1, '1', 2) : Array[AnyRef]`                   |
+| `toIArray` | `(1, 'a', 2).toIArray`                                      | `IArray(1, '1', 2) : IArray[AnyRef]`                 |
+| `map`      | `(1, 'a').map[[X] =>> Option[X]]([T] => (t: T) => Some(t))` | `(Some(1), Some('a')) : (Option[Int], Option[Char])` |
+
+
+# Under the hood: Scala 3 introduces match types
```

Comment:
I feel that this section could be a separate article. What you wrote above is already pretty dense :smile: 

At the same time, it makes me wonder… Do we have any guarantees that the two given instances you wrote for the transformation are total? Do we cover all the cases? Instead, could we implement it as a single recursive given instance and a match type (so that we would benefit from the exhaustivity checking of the compiler)?

Reply:
I did not think about splitting the article.
I think both "halves" are short enough and are closely linked.

But it is an appropriate remark, what do the others think ?

Reply:
I share your doubt about totality. The only case that does not seem to be covered to my eyes is `EmptyTuple`, for which we can add a simple encoder creating an empty list, but it is not that exciting.

What I can also do is having all the cases and maybe the type alias and split row/column encoders in a repo or gist that is linked in the article for completeness.
The idea would be to show the simplest and shortest approach in the article but provide a place to find a more complete solution. What do you think?

Reply:
> I did not think about splitting the article.
> I think both "halves" are short enough and are closely linked.
> 
> But it is an appropriate remark, what do the others think ?

I don't have a strong opinion, but I would slightly lean towards splitting into two articles.

Reply:
I thought a bit about it, but I prefer to keep the content together and not to split it in two articles.
The two subsections are linked and I believe that separating them will make each part look less "impressive".